File: workers/go/projects/harness/harness.go
package harness

import (
	"context"
	"crypto/tls"
	"errors"
	"flag"
	"fmt"
	"log"
	"net"
	"os"
	"strings"

	"github.com/temporalio/omes/workers/go/projects/api"
	"go.temporal.io/api/enums/v1"
	"go.temporal.io/api/operatorservice/v1"
	"go.temporal.io/sdk/client"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

const (
	OmesSearchAttributeKey = "OmesExecutionID"
	defaultHost            = "0.0.0.0"
)

// ClientFunc creates a Temporal client from the given options. Called by both
// the worker subprocess and the project-server Init handler.
type ClientFunc func(client.Options, ClientConfig) (client.Client, error)

// ClientConfig is passed to ClientFunc with minimal connection-time data.
type ClientConfig struct {
	TaskQueue string
}

// InitFunc performs project-specific setup after the client is created.
// Only called by the project-server Init gRPC handler, not by the worker.
type InitFunc func(context.Context, client.Client, InitConfig) error

// InitConfig is passed to InitFunc with run-level configuration.
type InitConfig struct {
	ConfigJSON  []byte // from --config-file, project-specific schema
	TaskQueue   string
	RunID       string
	ExecutionID string
}

// ExecuteFunc is called once per Execute RPC (one per iteration).
type ExecuteFunc func(context.Context, client.Client, ExecuteInfo) error

// ExecuteInfo is passed to ExecuteFunc for each iteration.
type ExecuteInfo struct {
	TaskQueue   string
	ExecutionID string
	RunID       string
	Iteration   int64
	Payload     []byte // from executor, executor-specific schema
}

// WorkerFunc runs the worker.
type WorkerFunc func(client.Client, WorkerConfig) error

// WorkerConfig is passed to WorkerFunc.
type WorkerConfig struct {
	TaskQueue         string
	PromListenAddress string
}

// Harness bridges the omes CLI with project test code.
type Harness struct {
	runCtx      *runContext
	client      client.Client
	clientFunc  ClientFunc
	initFunc    InitFunc
	workerFunc  WorkerFunc
	executeFunc ExecuteFunc
	api.UnsafeProjectServiceServer
}

func New() *Harness {
	return &Harness{}
}

func (h *Harness) RegisterClient(fn ClientFunc) {
	if h.clientFunc != nil {
		log.Fatalf("client factory already registered")
	}
	h.clientFunc = fn
}

func (h *Harness) OnInit(fn InitFunc) {
	if h.initFunc != nil {
		log.Fatalf("init handler already registered")
	}
	h.initFunc = fn
}

func (h *Harness) RegisterWorker(fn WorkerFunc) {
	if h.workerFunc != nil {
		log.Fatalf("worker already registered")
	}
	h.workerFunc = fn
}

func (h *Harness) OnExecute(fn ExecuteFunc) {
	if h.executeFunc != nil {
		log.Fatalf("execute handler already registered")
	}
	h.executeFunc = fn
}

func (h *Harness) Run() {
	if h.clientFunc == nil {
		log.Fatalf("no client factory registered; call RegisterClient before Run")
	}
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "usage: <program> [project-server|worker] ...")
		os.Exit(1)
	}
	cmd := os.Args[1]
	os.Args = append(os.Args[:1], os.Args[2:]...)

	switch cmd {
	case "project-server":
		h.startGrpcServer()
	case "worker":
		h.startWorker()
	default:
		fmt.Fprintf(os.Stderr, "unknown command: %s\n", cmd)
		os.Exit(1)
	}
}

// Init handles the Init gRPC call from the omes CLI.
func (s *Harness) Init(ctx context.Context, req *api.InitRequest) (*api.InitResponse, error) {
	if err := validateInit(req); err != nil {
		return nil, err
	}

	co := req.GetConnectOptions()
	opts, err := buildClientOptions(
		co.GetServerAddress(),
		co.GetNamespace(),
		co.GetAuthHeader(),
		tlsOptions{
			EnableTLS:               co.GetEnableTls(),
			CertPath:                co.GetTlsCertPath(),
			KeyPath:                 co.GetTlsKeyPath(),
			ServerName:              co.GetTlsServerName(),
			DisableHostVerification: co.GetDisableHostVerification(),
		},
	)
	if err != nil {
		return nil, status.Error(codes.InvalidArgument, err.Error())
	}

	c, err := s.clientFunc(opts, ClientConfig{
		TaskQueue: req.GetTaskQueue(),
	})
	if err != nil {
		return nil, status.Error(codes.Internal, fmt.Sprintf("client creation failed: %v", err))
	}
	s.client = c

	if s.initFunc != nil {
		if err := s.initFunc(ctx, c, InitConfig{
			ConfigJSON:  req.GetConfigJson(),
			TaskQueue:   req.GetTaskQueue(),
			RunID:       req.GetRunId(),
			ExecutionID: req.GetExecutionId(),
		}); err != nil {
			return nil, status.Error(codes.Internal, fmt.Sprintf("init failed: %v", err))
		}
	}

	if req.RegisterSearchAttributes {
		if err := s.registerSearchAttributes(ctx, co.GetNamespace()); err != nil {
			return nil, status.Error(codes.FailedPrecondition, err.Error())
		}
	}

	s.runCtx = &runContext{
		TaskQueue:   req.GetTaskQueue(),
		ExecutionID: req.GetExecutionId(),
		RunID:       req.GetRunId(),
	}

	return &api.InitResponse{}, nil
}

// Execute handles the Execute gRPC call from the omes CLI.
func (s *Harness) Execute(ctx context.Context, req *api.ExecuteRequest) (*api.ExecuteResponse, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "execute request is nil")
	}
	if s.executeFunc == nil {
		return nil, status.Error(codes.FailedPrecondition, "no execute handler registered")
	}
	if s.client == nil {
		return nil, status.Error(codes.FailedPrecondition, "must initialize harness before calling execute")
	}

	if err := s.executeFunc(ctx, s.client, ExecuteInfo{
		TaskQueue:   s.runCtx.TaskQueue,
		ExecutionID: s.runCtx.ExecutionID,
		RunID:       s.runCtx.RunID,
		Iteration:   req.Iteration,
		Payload:     req.Payload,
	}); err != nil {
		return nil, status.Error(codes.Internal, err.Error())
	}
	return &api.ExecuteResponse{}, nil
}

// --- Internal helpers ---

type runContext struct {
	TaskQueue   string
	ExecutionID string
	RunID       string
}

func (h *Harness) startGrpcServer() {
	if h.executeFunc == nil {
		log.Fatalf("Attempted to start server, but no execute handler was registered")
	}
	if h.workerFunc == nil {
		log.Fatalf("Attempted to start server, but no worker handler was registered")
	}

	fs := flag.NewFlagSet("client", flag.ExitOnError)
	port := fs.Int("port", 8080, "HTTP port")

	if err := fs.Parse(os.Args[1:]); err != nil {
		log.Fatalf("Failed to parse flags: %v", err)
	}

	lis, err := net.Listen("tcp", fmt.Sprintf("%s:%d", defaultHost, *port))
	if err != nil {
		log.Fatalf("Failed to listen: %v", err)
	}
	grpcServer := grpc.NewServer()
	api.RegisterProjectServiceServer(grpcServer, h)
	if err := grpcServer.Serve(lis); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
		log.Fatalf("gRPC server failed: %v", err)
	}
}

func validateInit(req *api.InitRequest) error {
	if req == nil {
		return status.Error(codes.InvalidArgument, "request: required")
	}
	if req.GetExecutionId() == "" {
		return status.Error(codes.InvalidArgument, "execution_id: required")
	}
	if req.GetRunId() == "" {
		return status.Error(codes.InvalidArgument, "run_id: required")
	}
	if req.GetTaskQueue() == "" {
		return status.Error(codes.InvalidArgument, "task_queue: required")
	}
	co := req.GetConnectOptions()
	if co == nil {
		return status.Error(codes.InvalidArgument, "connect_options: required")
	}
	if co.GetNamespace() == "" {
		return status.Error(codes.InvalidArgument, "connect_options.namespace: required")
	}
	if co.GetServerAddress() == "" {
		return status.Error(codes.InvalidArgument, "connect_options.server_address: required")
	}
	return nil
}

type staticHeadersProvider map[string]string

func (s staticHeadersProvider) GetHeaders(context.Context) (map[string]string, error) {
	return s, nil
}

type tlsOptions struct {
	EnableTLS               bool
	CertPath                string
	KeyPath                 string
	ServerName              string
	DisableHostVerification bool
}

func buildClientOptions(
	serverAddress string,
	namespace string,
	authHeader string,
	tlsOpts tlsOptions,
) (client.Options, error) {
	opts := client.Options{
		HostPort:  serverAddress,
		Namespace: namespace,
	}
	if authHeader != "" {
		opts.HeadersProvider = staticHeadersProvider{
			"Authorization": authHeader,
		}
	}
	tlsCfg, err := loadTLSConfig(tlsOpts)
	if err != nil {
		return client.Options{}, fmt.Errorf("failed to load TLS config: %w", err)
	}
	if tlsCfg != nil {
		opts.ConnectionOptions.TLS = tlsCfg
	}
	return opts, nil
}

func loadTLSConfig(opts tlsOptions) (*tls.Config, error) {
	tlsConfig := &tls.Config{
		InsecureSkipVerify: opts.DisableHostVerification,
		ServerName:         opts.ServerName,
		MinVersion:         tls.VersionTLS13,
	}
	if opts.CertPath != "" {
		if opts.KeyPath == "" {
			return nil, errors.New("got TLS cert with no key")
		}
		cert, err := tls.LoadX509KeyPair(opts.CertPath, opts.KeyPath)
		if err != nil {
			return nil, fmt.Errorf("failed to load certs: %w", err)
		}
		tlsConfig.Certificates = []tls.Certificate{cert}
		return tlsConfig, nil
	} else if opts.KeyPath != "" {
		return nil, errors.New("got TLS key with no cert")
	}
	if opts.EnableTLS {
		return tlsConfig, nil
	}
	return nil, nil
}

func (s *Harness) registerSearchAttributes(ctx context.Context, namespace string) error {
	_, err := s.client.OperatorService().AddSearchAttributes(ctx, &operatorservice.AddSearchAttributesRequest{
		SearchAttributes: map[string]enums.IndexedValueType{
			"KS_Keyword":           enums.INDEXED_VALUE_TYPE_KEYWORD,
			"KS_Int":               enums.INDEXED_VALUE_TYPE_INT,
			OmesSearchAttributeKey: enums.INDEXED_VALUE_TYPE_KEYWORD,
		},
		Namespace: namespace,
	})
	if err != nil {
		if status.Code(err) == codes.AlreadyExists {
			return nil
		}
		if strings.Contains(err.Error(), "attributes mapping unavailable") {
			return nil
		}
		return fmt.Errorf("failed to register search attributes: %w", err)
	}
	return nil
}

File: workers/dotnet/projects/harness/Harness.cs
using System.CommandLine;
using System.CommandLine.Invocation;
using Google.Protobuf;
using Grpc.Core;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Server.Kestrel.Core;
using Microsoft.Extensions.DependencyInjection;
using Temporalio.Client;
using Temporal.Omes.Projects.V1;

namespace Temporalio.Omes.Projects.Harness;

/// <summary>
/// Minimal config passed to ClientFunc for client creation.
/// </summary>
public record ClientConfig(
    string TaskQueue
);

/// <summary>
/// Run-level config passed to InitFunc for project-specific setup.
/// </summary>
public record InitConfig(
    byte[]? ConfigJson,
    string TaskQueue,
    string RunId,
    string ExecutionId
);

/// <summary>
/// Per-iteration data passed to ExecuteFunc.
/// </summary>
public record ExecuteInfo(
    string TaskQueue,
    string ExecutionId,
    string RunId,
    long Iteration,
    byte[]? Payload
);

/// <summary>
/// Worker configuration passed to WorkerFunc.
/// </summary>
public record WorkerConfig(
    string TaskQueue,
    string? PromListenAddress = null
);

/// <summary>
/// Bridges the omes CLI with .NET project test code via gRPC.
/// </summary>
public partial class ProjectHarness
{
    public const string OmesSearchAttributeKey = "OmesExecutionID";

    public delegate Task<ITemporalClient> ClientFunc(
        TemporalClientConnectOptions opts, ClientConfig config);

    public delegate Task InitFunc(
        ITemporalClient client, InitConfig config);

    public delegate Task ExecuteFunc(
        ITemporalClient client, ExecuteInfo info);

    public delegate Task WorkerFunc(
        ITemporalClient client, WorkerConfig config);

    private ClientFunc? _clientFn;
    private InitFunc? _initFn;
    private WorkerFunc? _workerFn;
    private ExecuteFunc? _executeFn;
    private ITemporalClient? _client;
    private RunContext? _runCtx;

    public void RegisterClient(ClientFunc fn)
    {
        if (_clientFn != null)
            throw new InvalidOperationException("Client factory already registered");
        _clientFn = fn;
    }

    public void OnInit(InitFunc fn)
    {
        if (_initFn != null)
            throw new InvalidOperationException("Init handler already registered");
        _initFn = fn;
    }

    public void RegisterWorker(WorkerFunc fn)
    {
        if (_workerFn != null)
            throw new InvalidOperationException("Worker already registered");
        _workerFn = fn;
    }

    public void OnExecute(ExecuteFunc fn)
    {
        if (_executeFn != null)
            throw new InvalidOperationException("Execute handler already registered");
        _executeFn = fn;
    }

    public async Task<int> RunAsync(string[] args)
    {
        if (_clientFn == null)
            throw new InvalidOperationException("No client factory registered; call RegisterClient before RunAsync");

        var root = new RootCommand("Project test harness");
        root.Add(BuildWorkerCommand());

        // Project-server subcommand
        var portOption = new Option<int>("--port", () => 8080, "gRPC server port");
        var serverCmd = new Command("project-server", "Run the project gRPC server");
        serverCmd.Add(portOption);
        serverCmd.SetHandler(async (InvocationContext ctx) =>
        {
            if (_executeFn == null)
                throw new InvalidOperationException("No execute handler registered");
            if (_workerFn == null)
                throw new InvalidOperationException("No worker handler registered");

            var port = ctx.ParseResult.GetValueForOption(portOption);
            await StartGrpcServerAsync(port);
        });
        root.Add(serverCmd);

        return await root.InvokeAsync(args);
    }

    private async Task StartGrpcServerAsync(int port)
    {
        var builder = WebApplication.CreateBuilder();
        builder.WebHost.ConfigureKestrel(options =>
        {
            options.ListenAnyIP(port, listenOptions =>
            {
                listenOptions.Protocols = HttpProtocols.Http2;
            });
        });
        builder.Services.AddGrpc();
        builder.Services.AddSingleton(this);

        var app = builder.Build();
        app.MapGrpcService<ProjectServiceImpl>();

        await app.RunAsync();
    }

    internal async Task<InitResponse> HandleInit(InitRequest req)
    {
        ValidateInit(req);

        var co = req.ConnectOptions;
        var opts = await BuildClientConnectOptions(
            co.ServerAddress,
            co.Namespace,
            co.AuthHeader,
            co.EnableTls,
            co.TlsCertPath,
            co.TlsKeyPath,
            co.TlsServerName);

        _client = await _clientFn!(opts, new ClientConfig(req.TaskQueue));

        if (_initFn != null)
        {
            await _initFn(_client, new InitConfig(
                req.ConfigJson?.ToByteArray(),
                req.TaskQueue,
                req.RunId,
                req.ExecutionId
            ));
        }

        if (req.RegisterSearchAttributes)
        {
            await RegisterSearchAttributes(co.Namespace);
        }

        _runCtx = new RunContext(req.TaskQueue, req.ExecutionId, req.RunId);
        return new InitResponse();
    }

    internal async Task<ExecuteResponse> HandleExecute(ExecuteRequest req)
    {
        if (req == null)
            throw new RpcException(new Status(StatusCode.InvalidArgument, "execute request is nil"));
        if (_executeFn == null)
            throw new RpcException(new Status(StatusCode.FailedPrecondition, "no execute handler"));
        if (_runCtx == null)
            throw new RpcException(new Status(StatusCode.FailedPrecondition, "not initialized"));

        await _executeFn(_client!, new ExecuteInfo(
            _runCtx.TaskQueue,
            _runCtx.ExecutionId,
            _runCtx.RunId,
            req.Iteration,
            req.Payload?.ToByteArray()
        ));

        return new ExecuteResponse();
    }

    // --- Internal helpers ---

    private static void ValidateInit(InitRequest req)
    {
        if (req == null)
            throw new RpcException(new Status(StatusCode.InvalidArgument, "request: required"));
        if (string.IsNullOrEmpty(req.ExecutionId))
            throw new RpcException(new Status(StatusCode.InvalidArgument, "execution_id: required"));
        if (string.IsNullOrEmpty(req.RunId))
            throw new RpcException(new Status(StatusCode.InvalidArgument, "run_id: required"));
        if (string.IsNullOrEmpty(req.TaskQueue))
            throw new RpcException(new Status(StatusCode.InvalidArgument, "task_queue: required"));
        var co = req.ConnectOptions;
        if (co == null)
            throw new RpcException(new Status(StatusCode.InvalidArgument, "connect_options: required"));
        if (string.IsNullOrEmpty(co.Namespace))
            throw new RpcException(new Status(StatusCode.InvalidArgument, "connect_options.namespace: required"));
        if (string.IsNullOrEmpty(co.ServerAddress))
            throw new RpcException(new Status(StatusCode.InvalidArgument, "connect_options.server_address: required"));
    }

    internal static async Task<TemporalClientConnectOptions> BuildClientConnectOptions(
        string serverAddress,
        string ns,
        string? authHeader,
        bool enableTls,
        string? tlsCertPath,
        string? tlsKeyPath,
        string? tlsServerName)
    {
        var opts = new TemporalClientConnectOptions(serverAddress) { Namespace = ns };

        if (!string.IsNullOrEmpty(authHeader))
        {
            opts.RpcMetadata = new Dictionary<string, string>
            {
                ["Authorization"] = authHeader
            };
        }

        if (enableTls || !string.IsNullOrEmpty(tlsCertPath))
        {
            opts.Tls = new TlsOptions();
            if (!string.IsNullOrEmpty(tlsServerName))
            {
                opts.Tls.Domain = tlsServerName;
            }
            if (!string.IsNullOrEmpty(tlsCertPath) && !string.IsNullOrEmpty(tlsKeyPath))
            {
                opts.Tls.ClientCert = await File.ReadAllBytesAsync(tlsCertPath);
                opts.Tls.ClientPrivateKey = await File.ReadAllBytesAsync(tlsKeyPath);
            }
        }

        return opts;
    }

    private async Task RegisterSearchAttributes(string ns)
    {
        try
        {
            await _client!.Connection.OperatorService.AddSearchAttributesAsync(
                new Temporalio.Api.OperatorService.V1.AddSearchAttributesRequest
                {
                    Namespace = ns,
                    SearchAttributes =
                    {
                        { "KS_Keyword", Temporalio.Api.Enums.V1.IndexedValueType.Keyword },
                        { "KS_Int", Temporalio.Api.Enums.V1.IndexedValueType.Int },
                        { OmesSearchAttributeKey, Temporalio.Api.Enums.V1.IndexedValueType.Keyword },
                    }
                });
        }
        catch (RpcException ex) when (ex.StatusCode == StatusCode.AlreadyExists)
        {
            // Already registered, that's fine
        }
        catch (RpcException ex) when (ex.Message.Contains("attributes mapping unavailable"))
        {
            // Also fine
        }
    }

    private record RunContext(string TaskQueue, string ExecutionId, string RunId);
}

/// <summary>
/// gRPC service implementation that delegates to the harness.
/// </summary>
public class ProjectServiceImpl : ProjectService.ProjectServiceBase
{
    private readonly ProjectHarness _harness;

    public ProjectServiceImpl(ProjectHarness harness)
    {
        _harness = harness;
    }

    public override async Task<InitResponse> Init(InitRequest request, ServerCallContext context)
    {
        return await _harness.HandleInit(request);
    }

    public override async Task<ExecuteResponse> Execute(ExecuteRequest request, ServerCallContext context)
    {
        return await _harness.HandleExecute(request);
    }
}
